teps/0106-support-specifying-metadata-per-task-in-runtime.md

```diff
@@ -0,0 +1,209 @@
+
+## Design Details
+
+Guided by the stated “Reusability” by [Tekton Design Principles](https://github.com/tektoncd/community/blob/main/design-principles.md), the added metadata will be located under the `taskRunSpecs` / `spec` field of the `PipelineRun` type. This will allow a user specify more execution-context-related metadata in `PipelineRun` rather than being limited by a static definition for a `Task`.
+
+So the metadata field will be added as (addition marked with +):
+
+```go
+// PipelineTaskRunSpec an be used to configure specific
+// specs for a concrete Task
+type PipelineTaskRunSpec struct {
+        PipelineTaskName       string                   json:"pipelineTaskName,omitempty"
+        TaskServiceAccountName string                   json:"taskServiceAccountName,omitempty"
+        TaskPodTemplate        *PodTemplate             json:"taskPodTemplate,omitempty"
+        StepOverrides          []TaskRunStepOverride    json:"stepOverrides,omitempty"
+        SidecarOverrides       []TaskRunSidecarOverride json:"sidecarOverrides,omitempty"
+
++       // +optional
++       Metadata PipelineTaskMetadata json:"metadata,omitempty"
+}
+```
+
+And the referenced metadata type is defined as:
+
+```go
+// PipelineTaskMetadata contains the labels or annotations
+type PipelineTaskMetadata struct {
+        // +optional
+        Labels map[string]string json:"labels,omitempty"
+
+        // +optional
+        Annotations map[string]string json:"annotations,omitempty"
+}
+```
+
+An `PipelineRun` example to show the structure (addition marked with +):
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: ${pipeline-run-name} 
+spec:
+  pipelineRef:
+    name: ${pipeline-name}
+  taskRunSpecs:
+  - pipelineTaskName: ${task-name}
++   metadata: 
++     annotations:
++       vault.hashicorp.com/agent-inject-secret-${unique-name}: ${/path/to/secret}
++       vault.hashicorp.com/role: ${role}
+```
+
+## Design Evaluation
+
+This work is a user-facing change on API while following the `Reusability` principle to keep a loose coupling between a `Task` and a `TaskRun` definition.
+
+### Reusability
+
+A `Task` is expected to be better reused and keep flexibility while the runtime-related metadata can be independently added.  
+
+### Simplicity
+
+With this work, a user is expected to avoid defining multi `Task`s which only differentiate on certain required metadata in the runtime.
+
+### Drawbacks
+
+- [ ] #TODO: to collect from feedback
+
+## Alternatives
+
+### Add Metadata under `TaskRef` in `Pipeline`
+
+While referring to the [`EmbeddedTask`](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#EmbeddedTask) type under the `Pipeline` type, a possible solution will be adding a metadata field under the `TaskRef` as discussed [here](https://github.com/tektoncd/pipeline/issues/4105#issuecomment-1075335509).
```

Comment:
I think this is a viable alternative for certain scenarios e.g. the hermetic use case mentioned above. When authoring a pipeline, a user could want a way to indicate that some of the tasks should always run hermetically. That being said, I'm fine to track this is as a separate feature.

Reply:
thanks for sharing this thought! let me add as a note.